jax/_src/api.py

```diff
@@ -1302,13 +1302,17 @@ def _split(x, indices, axis):
     return x.split(indices, axis)
 
 
-def vmap(fun: F, in_axes=0, out_axes=0, axis_name=None, axis_size=None) -> F:
+def vmap(fun: F,
+         in_axes: Union[None, int, Tuple[Any, ...]] = 0,
```

Comment:
Since `in_axes` can be an arbitrary pytree, I think the only viable type annotation is `Any`.

Reply:
@jakevdp Hmmm, I just tested it:
```python
from jax import vmap
import jax.numpy as jnp

def f(d):
    return d['blah']


vf = vmap(f, in_axes={'blah': 0})

try:
    vf({'blah': jnp.zeros(10)})
except:
    print("guess not?")

vf_tuple = vmap(f, in_axes=({'blah': 0},))
print(vf_tuple({'blah': jnp.zeros(10)}))
```
Am I doing something wrong here?  It seems that `in_axes` should be a tuple corresponding to the parameters of `f`?

Reply:
I guess you're right, it needs to be a tuple of pytrees.

Reply:
Great :smile: Please let me know if the text can be improved at all.  I tried to keep it understandable for the most common cases, but it's hard when the underlying library is so powerful!

Reply:
sorry, I'm just overly suspicious of type annotation PRs because annotations tend to have very poor test coverage in JAX, and these types of PRs have often needed to be rolled-back due to unexpected failures in downstream packages.

Reply:
The benefit of `Any` is that you never have false positives 😀

Reply:
Looks like the lint check caught an issue with the `in_axes` annotation